core/include/assert.h

```diff
@@ -68,8 +61,11 @@ extern const char assert_crash_message[];
  * @param[in] file  The file name of the file the assertion failed in
  * @param[in] line  The code line of @p file the assertion failed in
  */
-NORETURN void _assert_failure(const char *file, unsigned line);
-
+#ifdef __GNUC__
+__attribute__((noreturn)) void _assert_failure(const char *file, unsigned line);
+#else
+void _assert_failure(const char *file, unsigned line);
```

Comment:
For these alone it may be convenient to rename NORETURN to __NORETURN (or _NORETURN, is there a difference?) and move it to a header that can be included by standard header files; the existing NORETURN could then be just a compatibility alias.

Reply:
`Also reserved for the implementor are all external identifiers beginning with an underscore, and
all other identifiers beginning with an underscore followed by a capital letter or an underscore.` (same doc same §)

these buildins kind of things of gcc are `__` : `__attribute__` `__LINE__`
glibc assert.h uses `__assert_fail`

Reply:
For example, cmsis_gcc.h defines
```
#define __NO_RETURN                            __attribute__((__noreturn__))
```

In general, this affects assert.h being a POSIX header, but also other headers we (have to) pull into projects that didn't ask for them some of the defines, like lwip's arch/cc.h. On the long run, I think it'd be good to have an alternative to (subset of) kernel_defines.h that is safe to use in non-RIOT headers (eg. defining __NORETURN rather than NORETURN).

But that's ambitious, and may be out of scope for this. If nothing else, this line should explain why NORETURN is not used here, and maybe reference a roadmap-issue on how to make it better.